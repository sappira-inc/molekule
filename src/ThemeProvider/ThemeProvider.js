import React from 'react';
import { ThemeProvider as SCThemeProvider } from 'styled-components';
import merge from 'lodash/merge';
import defaultTheme from '../theme';

const ThemeProvider = props => {
  const theme = merge(defaultTheme(props.theme), props.theme);

  return <SCThemeProvider theme={theme}>{props.children}</SCThemeProvider>;
};

export default ThemeProvider;
