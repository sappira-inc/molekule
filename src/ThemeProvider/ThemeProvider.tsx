import React, { FC } from 'react';
import { ThemeProvider as SCThemeProvider, DefaultTheme } from 'styled-components';
import merge from 'lodash/merge';
import { createTheme } from '../createTheme';

interface ThemeProviderProps {
  theme?: Partial<DefaultTheme>;
}

export const ThemeProvider: FC<ThemeProviderProps> = props => {
  const theme = merge(createTheme(props.theme), props.theme);

  return <SCThemeProvider theme={theme}>{props.children}</SCThemeProvider>;
};
