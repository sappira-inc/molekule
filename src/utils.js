import { get, kebabCase } from 'lodash';
import styled from 'styled-components';

export const themeGet = (theme, lookup, fallback) => get(theme, lookup, fallback);

export const getComponentVariant = (theme, componentName, variant) => {
  const config = themeGet(
    theme,
    `components.${componentName}.variants.${variant}`,
    theme.variants[componentName][variant]
  );
  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentStyle = componentName => ({ theme }) => themeGet(theme, `components.${componentName}.style`);

export const getComponentClassName = ({ className, theme: { classPrefix }, variant }, name) =>
  `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

export const createComponent = ({ name, tag = 'div', as, style, props: defaultProps = () => ({}) }) => {
  const component = as ? styled(as) : styled[tag];

  return component.attrs(props => ({
    ...defaultProps(props),
    ...props,
    className: getComponentClassName(props, kebabCase(name)),
  }))`
    ${style}
    ${getComponentStyle(name)}
    ${({ styles = {} }) => styles[name] || {}}
  `;
};
