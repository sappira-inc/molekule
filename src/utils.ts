import kebabCase from 'lodash/kebabCase';
import get from 'lodash/get';
import styled, { CSSProp } from 'styled-components';

export const themeGet = (lookup: any, fallback?: any) => ({ theme }: any = {}) => get(theme, lookup, fallback);

export const getComponentVariant = (theme: any, componentName: string, variant: string) => {
  const config = themeGet(`variants.${componentName}.${variant}`)({
    theme,
  });

  if (!config) throw new Error(`Molekule: "${variant}" variant not found in theme...`);
  return config;
};

export const getComponentSize = (theme: any, componentName: string, size: string) =>
  themeGet(`sizes.${componentName}.${size}`, {})({ theme });

export const getComponentStyle = (componentName: string) => themeGet(`styles.${componentName}`, {});

export const getVariantStyles = (componentName: string, variant: string) =>
  themeGet(`variants.${componentName}.${variant}.style`, {});

interface GetComponentClassNameProps {
  className: string;
  theme: {
    classPrefix: string;
  };
  variant: string;
}

const getComponentClassName = (
  { className, theme: { classPrefix }, variant }: GetComponentClassNameProps,
  name: string
) => `${className || ''} ${classPrefix}-${name} ${variant ? `${classPrefix}-${name}-${variant}` : ''}`.trim();

interface CreateComponentProps<T> {
  name: string;
  tag?: keyof JSX.IntrinsicElements | React.JSXElementConstructor<any>;
  as?: React.ComponentType<any>;
  style?: ((props: any) => CSSProp) | CSSProp;
  props?: (props: T) => any;
}

export const createComponent = <
  T extends object,
  O extends keyof JSX.IntrinsicElements | React.ComponentType<any> = 'div'
>({
  name,
  tag = 'div',
  as,
  style,
  props: getBaseProps = () => ({}),
}: CreateComponentProps<T>) => {
  const component = styled<O>((as || tag) as any);

  return component.attrs<T>((props: any) => {
    const baseProps = getBaseProps(props);
    const finalProps = {
      ...baseProps,
      ...props,
      className: baseProps.className || props.className,
    };

    return {
      ...finalProps,
      className: getComponentClassName(finalProps, kebabCase(name)),
    };
  })<T>`
  ${style}
  ${getComponentStyle(name)}
  ${(p: any) => getVariantStyles(name, p.variant)}
  ${({ styles = {} }) => styles[name] || {}}
`;
};

// eslint-disable-next-line
const focusableFilter = (node: any) =>
  node.tabIndex >= 0 && !node.hasAttribute('disabled') ? NodeFilter.FILTER_ACCEPT : NodeFilter.FILTER_SKIP;

type FilterFn = (node?: any) => any;

const createTreeWalker = (root: any, currentNode: any, filterFn: FilterFn = () => NodeFilter.FILTER_ACCEPT) => {
  const treeWalker = document.createTreeWalker(
    root,
    NodeFilter.SHOW_ELEMENT,
    {
      acceptNode: filterFn,
    },
    false
  );
  treeWalker.currentNode = currentNode;
  return treeWalker;
};

export const findPreviousFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.previousNode() || treeWalker.currentNode === root) {
    return treeWalker.lastChild();
  }

  return treeWalker.currentNode;
};

export const findNextFocusableElement = (root: any, currentNode: any) => {
  const treeWalker = createTreeWalker(root, currentNode, focusableFilter);

  if (!treeWalker.nextNode() || treeWalker.currentNode === root) {
    treeWalker.currentNode = root;
    return treeWalker.firstChild();
  }

  return treeWalker.currentNode;
};

export const getNextCursorPosition = (cursorPos: any, oldValue: any, newValue: any) => {
  const cursorDiff = newValue.length - oldValue.length;
  let nextPosition = cursorPos;

  if (cursorDiff > 1 || cursorDiff < -1) {
    nextPosition += cursorDiff < 0 ? cursorDiff + 1 : cursorDiff - 1;
  }

  return Math.max(nextPosition, 0);
};
