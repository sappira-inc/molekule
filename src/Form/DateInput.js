import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import DateFormatter from 'cleave.js/src/shortcuts/DateFormatter';
import Input from './Input';
import { createEasyInput } from './EasyInput';
import { getNextCursorPosition } from '../utils';

export const getRawMaxLength = pattern => {
  const formatter = new DateFormatter(pattern);
  const blocks = formatter.getBlocks();
  return blocks.reduce((sum, block) => sum + block, 0);
};

const formatDate = (pattern, delimiter, dateString = '') => {
  const formatter = new DateFormatter(pattern);

  // Process our date string, bounding values between 1 and 31, and prepending 0s for
  // for single digit blocks that can't have 2 numbers, e.g. 5
  let tmpDate = formatter.getValidatedDate(`${dateString}`);

  // Blocks look something like [2, 2, 4], telling us how long each chunk should be
  return formatter.getBlocks().reduce((str, blockLength, index, blockArr) => {
    const block = tmpDate.substring(0, blockLength);
    if (!block) {
      return str;
    }

    tmpDate = tmpDate.substring(blockLength);

    // Append the delimiter if our block is complete and we're not at the last block
    const shouldAppendDelimiter = block.length === blockLength && index < blockArr.length - 1;

    return `${str}${block}${shouldAppendDelimiter ? delimiter : ''}`;
  }, '');
};

function DateInput({ delimiter, pattern, forwardedRef, value: propValue, onKeyDown, onChange, ...inputProps }) {
  const format = value => formatDate(pattern, delimiter, value);
  const [currentValue, setValue] = useState(format(propValue));
  const inputRef = forwardedRef || useRef();

  useEffect(() => {
    if (propValue !== currentValue) {
      setValue(format(propValue));
    }
  }, [propValue]);

  const handleKeyDown = event => {
    const isLetterLike = /^\w{1}$/.test(event.key);
    if (isLetterLike && currentValue.replace(/\D/g, '').length >= getRawMaxLength(pattern)) {
      event.preventDefault();
      event.stopPropagation();
    }

    if (onKeyDown) {
      onKeyDown(event);
    }
  };

  const handleChange = (name, newValue, event) => {
    const nextValue = newValue.length < currentValue.length ? newValue.trim() : format(newValue);
    const nextCursorPosition = getNextCursorPosition(event.target.selectionStart, currentValue, nextValue);

    setValue(nextValue);
    setTimeout(() => {
      inputRef.current.setSelectionRange(nextCursorPosition, nextCursorPosition);
    });

    if (onChange) {
      onChange(name, nextValue, event);
    }
  };

  return (
    <Input
      forwardedRef={inputRef}
      value={currentValue}
      onKeyDown={handleKeyDown}
      onChange={handleChange}
      {...inputProps}
    />
  );
}

DateInput.propTypes = {
  ...Input.propTypes,
  pattern: PropTypes.arrayOf(PropTypes.string),
  delimiter: PropTypes.string,
};

DateInput.defaultProps = {
  pattern: ['m', 'd', 'Y'],
  delimiter: '/',
};

export default createEasyInput(DateInput);
