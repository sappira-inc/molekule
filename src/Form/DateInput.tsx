import React, { useState, useEffect, useRef } from 'react';
import DateFormatter from 'cleave.js/src/shortcuts/DateFormatter';
import { Input, InputProps } from './Input';
import { createEasyInput } from './EasyInput';
import { getNextCursorPosition } from '../utils';

export const getRawMaxLength = (pattern: any) => {
  const formatter = new DateFormatter(pattern, '1900-01-01', '2099-12-31');
  const blocks: any[] = formatter.getBlocks();
  return blocks.reduce((sum, block) => sum + block, 0);
};

const formatDate = (pattern: any, delimiter: any, dateString = '') => {
  const formatter = new DateFormatter(pattern, '1900-01-01', '2099-12-31');

  // Process our date string, bounding values between 1 and 31, and prepending 0s for
  // for single digit blocks that can't have 2 numbers, e.g. 5
  let tmpDate = formatter.getValidatedDate(`${dateString}`);

  // Blocks look something like [2, 2, 4], telling us how long each chunk should be
  return (formatter.getBlocks() as any[]).reduce((str, blockLength, index, blockArr) => {
    const block = tmpDate.substring(0, blockLength);
    if (!block) {
      return str;
    }

    tmpDate = tmpDate.substring(blockLength);

    // Append the delimiter if our block is complete and we're not at the last block
    const shouldAppendDelimiter = block.length === blockLength && index < blockArr.length - 1;

    return `${str}${block}${shouldAppendDelimiter ? delimiter : ''}`;
  }, '');
};

export interface DateInputProps extends InputProps {
  initialValue?: string;
  pattern?: string[];
  delimiter?: string;
}

export const DateInput: React.FC<DateInputProps & { children?: never }> = ({
  delimiter,
  pattern,
  forwardedRef,
  initialValue,
  value: propValue,
  onKeyDown,
  onChange,
  ...inputProps
}) => {
  const format = (value: any) => formatDate(pattern, delimiter, value);
  const [currentValue, setValue] = useState(initialValue || format(propValue));
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const inputRef = forwardedRef || useRef();
  const previousValue = useRef(propValue);

  useEffect(() => {
    if (previousValue.current !== propValue && propValue !== currentValue) {
      setValue(format(propValue));
    }
    previousValue.current = propValue;
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [propValue]);

  const handleKeyDown = (event: any) => {
    const isLetterLike = /^\w{1}$/.test(event.key);
    if (isLetterLike && currentValue.replace(/\D/g, '').length >= getRawMaxLength(pattern)) {
      event.preventDefault();
      event.stopPropagation();
    }

    if (onKeyDown) {
      onKeyDown(event);
    }
  };

  const handleChange = (name: any, newValue: any, event: any) => {
    const nextValue = newValue.length < currentValue.length ? newValue.trim() : format(newValue);
    const nextCursorPosition = getNextCursorPosition(event.target.selectionStart, currentValue, nextValue);

    setValue(nextValue);
    setTimeout(() => {
      inputRef.current.setSelectionRange(nextCursorPosition, nextCursorPosition);
    });

    if (onChange) {
      onChange(name, nextValue, event);
    }
  };

  return (
    <Input
      forwardedRef={inputRef}
      value={currentValue}
      onKeyDown={handleKeyDown}
      onChange={handleChange}
      {...inputProps}
    />
  );
};

DateInput.defaultProps = {
  ...Input.defaultProps,
  pattern: ['m', 'd', 'Y'],
  delimiter: '/',
};

export default createEasyInput(DateInput);
