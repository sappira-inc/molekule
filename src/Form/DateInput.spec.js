import React from 'react';
import { renderWithTheme, fireEvent, act } from '../../test/utils';
import DateInput, { getRawMaxLength } from './DateInput';
import ThemeProvider from '../ThemeProvider';

describe('<DateInput />', () => {
  const renderInput = props => {
    const utils = renderWithTheme(<DateInput placeholder="Input" {...props} />);
    return {
      ...utils,
      input: utils.getByPlaceholderText('Input'),
    };
  };

  const updateInputValue = (input, value) => {
    act(() => {
      fireEvent.change(input, { target: { value } });
    });
  };

  test('#getMaxDateLength', () => {
    expect(getRawMaxLength(['m', 'd', 'Y'])).toEqual(8);
    expect(getRawMaxLength(['m', 'd', 'y'])).toEqual(6);
  });

  test('snapshot', () => {
    const { asFragment } = renderInput();
    expect(asFragment()).toMatchSnapshot();
  });

  test('initally sets the value', () => {
    const value = '12/05/1992';
    const { input } = renderInput({ value });

    expect(input.value).toEqual(value);
  });

  test("prefixes month and day block with a 0 when second number isn't possible", () => {
    const { input } = renderInput();

    updateInputValue(input, '4');
    expect(input.value).toEqual('04/');

    updateInputValue(input, '04/4');
    expect(input.value).toEqual('04/04/');
  });

  test("doesn't prefix when second number is possible", () => {
    const { input } = renderInput();

    updateInputValue(input, '1');
    expect(input.value).toEqual('1');

    updateInputValue(input, '12/1');
    expect(input.value).toEqual('12/1');
  });

  test('truncates a date string that is too long', () => {
    const { input } = renderInput({ value: '12/05/19922222' });

    expect(input.value).toEqual('12/05/1992');
  });

  test('can remove trailing slash when backspacing', () => {
    const { input } = renderInput({ value: '12/05/' });

    fireEvent.keyDown(input, { key: 'Backspace' });
    updateInputValue(input, '12/05');
    expect(input.value).toEqual('12/05');
  });

  test('takes custom delimter and pattern', () => {
    const { input } = renderInput({ value: '2000-5', pattern: ['Y', 'm', 'd'], delimiter: '-' });

    expect(input.value).toEqual('2000-05-');
  });

  test('updates internally when value prop changes', () => {
    const { input, rerender } = renderInput({ value: '01/05' });
    rerender(
      <ThemeProvider>
        <DateInput placeholder="input" value="01/09/1990" />
      </ThemeProvider>
    );
    expect(input.value).toEqual('01/09/1990');
  });

  test('initially displays initialValue if provided', () => {
    const { input } = renderInput({ initialValue: '01/01/2000', value: '2000-01-01' });

    expect(input.value).toEqual('01/01/2000');
  });
});
