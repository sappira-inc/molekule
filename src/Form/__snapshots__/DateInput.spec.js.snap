// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DateInput /> snapshot 1`] = `
<DocumentFragment>
  .c1 {
  position: relative;
  -webkit-transition: 175ms;
  transition: 175ms;
}

.c1 + .c0 {
  margin-top: 1rem;
}

.c2 {
  position: relative;
}

.c3 {
  border: 1px solid #E8E9EA;
  height: 48px;
  display: block;
  outline: none;
  width: 100%;
  padding: 8px;
  border-radius: 8px;
  -webkit-transition: 250ms all;
  transition: 250ms all;
  -webkit-appearance: none;
  font-family: inherit;
  font-size: 14px;
  color: #494D55;
  box-sizing: border-box;
}

.c3:hover,
.c3:active {
  border-color: #A4A6AA;
}

.c3:focus {
  border-color: #226EFF;
}

.c3::-webkit-input-placeholder {
  color: #767980;
}

.c3::-moz-placeholder {
  color: #767980;
}

.c3:-ms-input-placeholder {
  color: #767980;
}

.c3::placeholder {
  color: #767980;
}

.c3[disabled] {
  background-color: #FFFFFF;
  border-color: #E8E9EA;
  color: #D1D2D5;
}

.c3[disabled]::-webkit-input-placeholder {
  color: #D1D2D5;
}

.c3[disabled]::-moz-placeholder {
  color: #D1D2D5;
}

.c3[disabled]:-ms-input-placeholder {
  color: #D1D2D5;
}

.c3[disabled]::placeholder {
  color: #D1D2D5;
}

<div
    class="c0 re-field c1"
  >
    <div
      class="re-input-container c2"
    >
      <input
        class="re-input c3"
        placeholder="Input"
        rows="3"
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
